mcs/class/referencesource/System/net/System/URI.cs

```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
         private string GetLocalPath(){
             EnsureParseRemaining();
 
+            // https://bugzilla.xamarin.com/show_bug.cgi?id=58400
+            bool isFileUrlWithHost = (
+                (m_Info.Offset.Host != m_Info.Offset.Path) && 
+                // only file URLs
+                IsFile && 
+                // Paths that got reinterpreted as files would be treated
+                //  as if they have a host, like "/a/b" which maps to "file:///a/b/",
+                //  or "C:\a\b\" which maps to "file:///c:/a/b"
+                OriginalString.StartsWith("file://") &&
+                // file://localhost/x/y needs to produce /x/y
+                !IsLoopback
+            );
+
+            // Manually check for "hosts" that are just forward slashes.
+            if (isFileUrlWithHost) {
+                isFileUrlWithHost = false;
+
+                for (int i = m_Info.Offset.Host; i < m_Info.Offset.Path; i++) {
+                    if (OriginalString[i] != '/') {
+                        isFileUrlWithHost = true;
+                        break;
+                    }
+                }
+            }
+
+            // We need to force generation of a UNC-style path instead of a unix one,
+            // despite the path not being UNC originally
+            bool treatAsUncPath = IsUncPath || isFileUrlWithHost;
+
 #if MONO
             //
             // I think this is wrong but it keeps LocalPath fully backward compatible
             //
-            if (IsUncOrDosPath && (IsWindowsFileSystem || !IsUncPath))
+            if (
```

Comment:
Remove unneeded parens and after that one line should be more readable

Reply:
The precedence seems super unclear and easy to break without them. I got it wrong the first time. Can change to one line if you want